samples/web/content/apprtc/gcm_register.py

```diff
@@ -0,0 +1,221 @@
+
+  def handle_query(self):
+    msg = self.get_message_from_json()
+    if has_msg_field(msg, PARAM_USER_ID_LIST) and \
+        isinstance(msg[PARAM_USER_ID_LIST], list):
+      result = []
+      for id in msg[PARAM_USER_ID_LIST]:
+        # TODO(jiayl): Only return the verified users when SMS verification is
+        # added.
+        if len(GCMRecord.get_by_user_id(id)) > 0:
+          result.append(id)
+      self.response.out.write(json.dumps(result))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_del(self):
+    msg = self.get_message_from_json()
+    if has_msg_fields(msg, [PARAM_USER_ID, PARAM_GCM_ID]):
+      GCMRecord.remove(msg[PARAM_USER_ID], msg[PARAM_GCM_ID])
+      self.response.out.write(constants.RESPONSE_SUCCESS)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def post(self, cmd):
+    if cmd == 'new':
+      self.handle_new()
+    elif cmd == 'update':
+      self.handle_update()
+    elif cmd == 'verify':
+      self.handle_verify()
+    elif cmd == 'del':
+      self.handle_del()
+    elif cmd == 'query':
+      self.handle_query()
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+class GCMRecord(ndb.Model):
+  """Models a user ID to GCM ID binding record."""
+  user_id = ndb.StringProperty()
+  gcm_id = ndb.StringProperty()
+  code = ndb.StringProperty()
+  code_sent_time = ndb.DateTimeProperty()
+  verified = ndb.BooleanProperty()
+  last_modified_time = ndb.DateTimeProperty()
+
+  @classmethod
+  def get_by_user_id(cls, user_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key())
+    return q.filter(GCMRecord.user_id == user_id).fetch()
+
+  @classmethod
+  def get_by_gcm_id(cls, gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key())
+    return q.filter(GCMRecord.gcm_id == gcm_id).fetch()
+
+  @classmethod
+  def get_by_ids(cls, user_id, gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
+    return q.fetch()
+
+  @classmethod
+  @ndb.transactional(retries=100)
+  def add_or_update(cls, user_id, gcm_id, code):
+    records = GCMRecord.get_by_ids(user_id, gcm_id)
+    now = datetime.datetime.utcnow()
+    if len(records) == 0:
+      record = GCMRecord(parent = get_ancestor_key(),
+                         user_id = user_id,
+                         gcm_id = gcm_id,
+                         code = code,
+                         verified = False,
+                         code_sent_time = now,
+                         last_modified_time = now)
+      record.put()
+      logging.info(
+          'GCM binding added, user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+      return constants.RESPONSE_CODE_SENT
+
+    assert len(records) == 1
+    if records[0].verified:
+      logging.warning('Cannot update GCM binding code since already verified, '\
```

Comment:
Won't we still need to update the GCM ID post-verification? (e.g. on app upgrade)